shared/src/ledger/storage/merkle_tree.rs

```diff
@@ -156,136 +253,142 @@ impl<H: StorageHasher + Default> MerkleTree<H> {
         }
     }
 
-    /// Get the existence proof
-    pub fn get_existence_proof(
+    /// Get the existence proof from a sub-tree
+    pub fn get_sub_tree_existence_proof(
         &self,
-        key: &Key,
-        value: Vec<u8>,
-    ) -> Result<Proof> {
-        let (store_type, sub_key) = StoreType::sub_key(key)?;
-        let subtree = self.tree(&store_type);
-
-        // Get a proof of the sub tree
-        let hashed_sub_key = H::hash(&sub_key.to_string());
-        let cp = subtree.membership_proof(&hashed_sub_key)?;
-        // Replace the values and the leaf op for the verification
-        let sub_proof = match cp.proof.expect("The proof should exist") {
-            Ics23Proof::Exist(ep) => CommitmentProof {
-                proof: Some(Ics23Proof::Exist(ExistenceProof {
-                    key: sub_key.to_string().as_bytes().to_vec(),
-                    value,
-                    leaf: Some(self.leaf_spec()),
-                    ..ep
-                })),
-            },
-            // the proof should have an ExistenceProof
-            _ => unreachable!(),
-        };
-        self.get_proof(key, sub_proof)
+        keys: &[Key],
+        values: Vec<MerkleValue>,
+    ) -> Result<MembershipProof> {
+        let first_key = keys.iter().next().ok_or_else(|| {
+            Error::InvalidMerkleKey(
+                "No keys provided for existence proof.".into(),
+            )
+        })?;
+        let (store_type, _) = StoreType::sub_key(first_key)?;
+        if !keys.iter().all(|k| {
+            if let Ok((s, _)) = StoreType::sub_key(k) {
+                s == store_type
+            } else {
+                false
+            }
+        }) {
+            return Err(Error::InvalidMerkleKey(
+                "Cannot construct inclusion proof for keys in separate \
+                 sub-trees."
+                    .into(),
+            ));
+        }
+        self.tree(&store_type)
+            .subtree_membership_proof(keys, values)
     }
 
     /// Get the non-existence proof
     pub fn get_non_existence_proof(&self, key: &Key) -> Result<Proof> {
         let (store_type, sub_key) = StoreType::sub_key(key)?;
-        let subtree = self.tree(&store_type);
+        if store_type != StoreType::Ibc {
+            return Err(Error::NonExistenceProof(store_type.to_string()));
+        }
+
+        let string_key =
+            StringKey::try_from_bytes(sub_key.to_string().as_bytes())?;
+        let mut nep = self.ibc.non_membership_proof(&string_key)?;
+        // Replace the values and the leaf op for the verification
+        if let Some(ref mut nep) = nep.proof {
+            match nep {
+                Ics23Proof::Nonexist(ref mut ep) => {
+                    let NonExistenceProof {
+                        ref mut left,
+                        ref mut right,
+                        ..
+                    } = ep;
+                    let ep = left.as_mut().or(right.as_mut()).expect(
```

Comment:
No, it should just modify left.

Reply:
I think we have to replace the leaf spec on the right too.
hmm, both specs don't require to be replaced because the spec of the arse-merkle-tree is the same as `ibc_leaf_spec` for now.

Reply:
I'll modify both. That way we should be future proof.